ZEP000-build-state-on-followers.md

```diff
@@ -0,0 +1,508 @@
+
+One important part before we start with the Follower "Replay Mode" is that we need to restore the current state from the latest snapshot. This is necessary because when we process commands we immediately apply the follow-up events to our state, which we write to the log later. We can't be sure what we have already applied is committed, and to not apply events twice it is easier to just restore from the latest snapshot, KISS principle. Other than that it is similar to described [above](#bootstrap-follower-partition).
+
+### Re-init Follower Partition
+
+This can be seen as a special case and can also be called as Follower-To-Follower transition. In RAFT it can happen that InstallRequests are send to an slow Follower or to one which is lagging behind. In order to get the Follower back to the latest state, the Leader will send the Follower his latest snapshot.
+
+The Follower will reset his log, such that no gaps in the log exist. The Leader will send log entries after the snapshot index to the follower only. In order to replay with the latest state and on the correct index and position it is necessary to completely reset the follower. The easiest way is to trigger a new follower transition, which is similar to [Bootstrap the Follower partition](#bootstrap-follower-partition). We just need to make sure that the received snapshot is copied into the runtime folder.
+
+  
+[comment]: <> (          We have several challenges we need to overcome to solve/implement build state on followers. Let's take a look of the following causality chain:)
+[comment]: <> (           * To have a working and long living system, we need to be able to compact our log no matter the role. )
+[comment]: <> (           * To compact our log, we need to be able to take snapshots from the state.)
+[comment]: <> (             * The state need to contain a dense reflection of the log, last exported and processed position, from which we can recover.)
+[comment]: <> (           * To take a snasphot from the state, we need to build the state on all participants &#40;leader and follower&#41; in order to not replicate it.)
+[comment]: <> (           * To build state on followers, we need to avoid creating follow-up events by the state machine, Single Writer Principle. )
+[comment]: <> (           * To avoid follow-up events, we need to either replay only or use a NoopWriter.)
+[comment]: <> (In the following sections we will go in more depth in some these mentioned parts of the chain.)
+
+### Compacting the log
+
+In order to avoid an ever-growing log we need to compact it from time to time. This can be done by taking snapshots from the state, which allows us to start from that specific point, such that we can compact our log. With building state on followers each node is responsible for taking their own snapshots and compacting their log.
```

Comment:
🔧  In the guidelevel explanation it makes sense to quickly mention what log compaction actually is. In addition, let's add what we can achieve with this.
```suggestion
In order to avoid an ever-growing log we need to compact it from time to time. This can be done by taking snapshots from the state, which allows us to start from that specific point and compact the log (i.e. delete exported records from the log preceding the snapshot). By building state on followers each node can be responsible for taking their own snapshots and compacting their log. This reduces network traffic, because snapshots no longer need to be replicated.
```